execution_engine/src/shared/gas.rs

```diff
@@ -40,170 +42,195 @@ impl fmt::Display for Gas {
     }
 }
 
-impl std::ops::Add for Gas {
-    type Output = Gas;
+impl CheckedAdd for Gas {
+    fn checked_add(&self, v: &Self) -> Option<Self> {
+        self.0.checked_add(v.0).map(Gas)
+    }
+}
 
-    fn add(self, rhs: Self) -> Self::Output {
-        let val = self.value() + rhs.value();
-        Gas::new(val)
+impl CheckedMul for Gas {
+    fn checked_mul(&self, v: &Self) -> Option<Self> {
+        self.0.checked_mul(v.0).map(Gas)
     }
 }
 
-impl std::ops::Sub for Gas {
-    type Output = Gas;
+impl CheckedSub for Gas {
+    fn checked_sub(&self, v: &Self) -> Option<Self> {
+        self.0.checked_sub(v.0).map(Gas)
+    }
+}
 
-    fn sub(self, rhs: Self) -> Self::Output {
-        let val = self.value() - rhs.value();
-        Gas::new(val)
+impl Sum for Gas {
+    fn sum<I: Iterator<Item = Self>>(iter: I) -> Self {
+        iter.fold(Gas::zero(), std::ops::Add::add)
     }
 }
 
-impl std::ops::Div for Gas {
-    type Output = Gas;
+impl Integer for Gas {
+    fn div_floor(&self, other: &Self) -> Self {
+        Gas(self.0.div_floor(&other.0))
+    }
 
-    fn div(self, rhs: Self) -> Self::Output {
-        let val = self.value() / rhs.value();
-        Gas::new(val)
+    fn mod_floor(&self, other: &Self) -> Self {
+        Gas(self.0.mod_floor(&other.0))
     }
-}
 
-impl std::ops::Mul for Gas {
-    type Output = Gas;
+    fn gcd(&self, other: &Self) -> Self {
+        Gas(self.0.gcd(&other.0))
+    }
 
-    fn mul(self, rhs: Self) -> Self::Output {
-        let val = self.value() * rhs.value();
-        Gas::new(val)
+    fn lcm(&self, other: &Self) -> Self {
+        Gas(self.0.lcm(&other.0))
     }
-}
 
-impl std::ops::AddAssign for Gas {
-    fn add_assign(&mut self, rhs: Self) {
-        self.0 += rhs.0
+    fn divides(&self, other: &Self) -> bool {
+        self.0.divides(&other.0)
     }
-}
 
-impl Zero for Gas {
-    fn zero() -> Self {
-        Gas::new(U512::zero())
+    fn is_multiple_of(&self, other: &Self) -> bool {
+        self.0.is_multiple_of(&other.0)
     }
 
-    fn is_zero(&self) -> bool {
-        self.0.is_zero()
+    fn is_even(&self) -> bool {
+        self.0.is_even()
     }
-}
 
-impl Sum for Gas {
-    fn sum<I: Iterator<Item = Self>>(iter: I) -> Self {
-        iter.fold(Gas::zero(), std::ops::Add::add)
+    fn is_odd(&self) -> bool {
+        self.0.is_odd()
+    }
+
+    fn div_rem(&self, other: &Self) -> (Self, Self) {
+        let (div, rem) = self.0.div_rem(&other.0);
+        (Gas(div), Gas(rem))
     }
 }
 
 impl From<u32> for Gas {
```

Comment:
When do we cast a `u32` into `Gas`?  Maybe we don't need this `impl`?

Reply:
All costs on our platform are represented in u32 (host functions, opcodes, storage, etc.). I think it's safe to implement this trait for any integer type T that's smaller or equal to u64 in size, then create a Gas instance from a specific cost and continue math using Gas objects.